source/Tokens/Literals/Number.ts

```diff
@@ -0,0 +1,164 @@
+						case "2": value += 2; original += reader.read(); break
+						case "3": value += 3; original += reader.read(); break
+						case "4": value += 4; original += reader.read(); break
+						case "5": value += 5; original += reader.read(); break
+						case "6": value += 6; original += reader.read(); break
+						case "7": value += 7; original += reader.read(); break
+						case "8": value += 8; original += reader.read(); break
+						case "9": value += 9; original += reader.read(); break
+						default: result = new Number(divisor ? value / divisor : value, original, reader.mark()); continue
+					}
+					value *= 10
+					if (divisor)
+						divisor *= 10
+				}
+			} else {
+				result = null
+			}
+			return result
+		}
+		private static isNumber(character: string): boolean {
+			return character >= "0" && character <= "9"
+		}
+		private static scanOctal(reader : IO.BufferedReader): Token {
+			var result: Token
+			if (reader.peek(2) == "0o") {
+				var original = reader.read(2)
+				var value: number
+				var divisor: number
+				while (!result) {
+					switch (reader.peek()) {
+						case ".":
+							original += reader.read()
+							if (divisor)
+								result = new Number(divisor ? value / divisor : value, original, reader.mark())
+							else
+								divisor = 1
+							continue
+						case "_": original += reader.read(); continue
+						case "0": value += 0; original += reader.read(); break
+						case "1": value += 1; original += reader.read(); break
+						case "2": value += 2; original += reader.read(); break
+						case "3": value += 3; original += reader.read(); break
+						case "4": value += 4; original += reader.read(); break
+						case "5": value += 5; original += reader.read(); break
+						case "6": value += 6; original += reader.read(); break
+						case "7": value += 7; original += reader.read(); break
+						default: result = new Number(divisor ? value / divisor : value, original, reader.mark()); continue
+					}
+					value *= 8
+					if (divisor)
+						divisor *= 8
+				}
+			} else {
+				result = null
+			}
+			return result
+		}
+		private static scanBinary(reader : IO.BufferedReader): Token {
+			var result: Token
+			if (reader.peek(2) == "0b") {
+				var original = reader.read(2)
+				var value: number
+				var divisor: number
+				while (!result) {
+					switch (reader.peek()) {
+						case ".":
+							original += reader.read()
+							if (divisor)
+								result = new Number(divisor ? value / divisor : value, original, reader.mark())
+							else
+								divisor = 1
+							continue
+						case "_": original += reader.read(); continue
+						case "0": value += 0; original += reader.read(); break
+						case "1": value += 1; original += reader.read(); break
+						default: result = new Number(divisor ? value / divisor : value, original, reader.mark()); continue
+					}
+					value *= 2
+					if (divisor)
+						divisor *= 2
+				}
+			} else {
+				result = null
+			}
+			return result
+		}
+	}
+}
```

Comment:
A solution such as this one (quick mock-up) is more elegant and easier to maintain in my opinion. For the sake of saving space, I've only included base 2, 8 and 16.

```Typescript
class NumericLiteral {
	static scan(literal: string): number {
		var result: number;
		switch (literal.substring(0, 2)) {
			case "0b":
				result = this.lexNumericLiteral(literal, 2, NumericLiteral.isBinaryDigit);
				break;
			case "0c":
				result = this.lexNumericLiteral(literal, 8, NumericLiteral.isOctalDigit);
				break;
			case "0x":
				result = this.lexNumericLiteral(literal, 16, NumericLiteral.isHexadecimalDigit);
				break;
			default:
				break;
		}
		return result;
	}
	private static lexNumericLiteral(literal: string, base: number, validator: (digit: string) => boolean): number {
		var index = 2; // simulate moving past the base specifier
		var result = "";
		while (validator(literal.charAt(index))) {
			result += literal.charAt(index++);
		}
		// original = literal.substring(0, 2) + result
		return parseInt(result, base);
	}
	private static isBinaryDigit(digit: string) {
		return digit >= "0" && digit <= "1";
	}
	private static isOctalDigit(digit: string) {
		return digit >= "0" && digit <= "7";
	}
	private static isHexadecimalDigit(digit: string) {
		return (digit >= "0" && digit <= "9") || (digit >= "a" && digit <= "f") || (digit >= "A" && digit <= "F");
	}
}
```
tested with
```Typescript
console.log(NumericLiteral.scan("0b10111101010001111") === 96911);
console.log(NumericLiteral.scan("0c34341327") === 7455447)
console.log(NumericLiteral.scan("0x167FD6") === 1474518);
console.log(NumericLiteral.scan("0b1LEXER_WILL_STOP_BEFORE_THIS_STRING") === 1);
console.log(NumericLiteral.scan("0c7fLEXER_WILL_STOP_BEFORE_THIS_STRING") === 7);
console.log(NumericLiteral.scan("0xfLEXER_WILL_STOP_BEFORE_THIS_STRING") === 15);
```
What are your thoughts about using something like this instead?